homework11/homework11_4.py

```diff
@@ -0,0 +1,49 @@
+"""Homework11_4:Decorators. Writing and implementing programs."""
+
+from typing import Dict
+
+# Функция кэширования *
+# Напишите декоратор @cache, который кэширует результаты вызова функции и
+# возвращает закэшированное значение при повторных вызовах с теми же
+# аргументами. Это поможет избежать повторных вычислений для одинаковых
+# входных данных и ускорит работу программы.
+# Подсказки:
+# Используйте словарь для хранения закэшированных значений. Ключом словаря
+# будет набор аргументов функции, а значением - результат вызова функции
+# с этими аргументами.
+# Внутри декоратора, передайте аргументы функции в качестве ключа для доступа
+# к закэшированным значениям.
+# Если ключ уже есть в словаре, верните соответствующее значение. Если ключа
+# нет, вызовите функцию, сохраните результат в словаре и верните его.
+
+cached_results: Dict[str, int] = {}
+
+
+def cache(func):
+    """ Decorator to cache the results of a function.
+    When the function is called with the same arguments,
+    it returns the cached result, avoiding repeated computations.
+   """
+
+    def wrapper(n):
+        if cached_results.get(n) is not None:
+            return cached_results[n]
+        result = func(n)
```

Comment:
It is better to define variable inside

```suggestion
    def wrapper(n):
        cached_results = {}
        if cached_results.get(n) is not None:
            return cached_results[n]
        result = func(n)
```

Reply:
Sorry, but its not clear, if I move cached_result inside wrapper , I have something like: {}
{}
{}
{}
But if I use cached_result as global variable I have: {}
{1: 1}
{1: 1, 0: 0, 2: 1}
{1: 1, 0: 0, 2: 1, 3: 2}.
Are you sure that is good idea to place it cached_result inside wrapper? If so, I will move it inside the wrapper. Sorry but I am a little confused with this task :(((

Reply:
My bad, it should be like this:
```
def cache(func):
    cached_results = {}  # словарь для хранения закэшированных результатов

    def wrapper(*args):
        if args in cached_results:  # проверяем, есть ли результат в кэше
            return cached_results[args]
        result = func(*args)  # вызываем функцию и сохраняем результат
        cached_results[args] = result  # сохраняем результат в кэше
        return result

    return wrapper
```